google/cloud/security/common/gcp_type/firewall_rule.py

```diff
@@ -65,11 +83,189 @@ def __init__(self, **kwargs):
         self.denied = parser.json_unstringify(
             kwargs.get('firewall_rule_denied'))
         if self.allowed and self.denied:
-            raise ValueError('Cannot have allowed and denied rules')
+            raise InvalidFirewallRuleError(
+                'Cannot have allowed and denied rules (%s, %s)' % (
+                    self.allowed, self.denied))
         if self.allowed is None and self.denied is None:
-            raise ValueError('Either allowed or denied rules must be set')
+            raise InvalidFirewallRuleError('Must have allowed or denied rules')
+        if parser.json_unstringify(kwargs.get('validate', 'false')):
```

Comment:
It seems messy to bundle validate into kwargs. This is a general issue of using kwargs for object initialization that should be addressed throughout the gcp_type library as part of the move to inventory v2 and auditor v2. For now I recommend explicitly moving validate to the __init__ arguments list with a default value of False.

Reply:
Fair point, done.